### PR TITLE
Introduce `WfsSearchInput` component

### PR DESCRIPTION
**Pull request by annarieger**
### Description:
Introduced `WfsSearchInput` component which can be actually regarded as an alternative to already existing `WfsSearch` component which depends on `ant` `Select``component.

The new component based on `ant` `Input` field only and shows by default no output on successfully fetched features. The way how retrieved features should be represented is determined from user and can be passed with  `onFetchSuccess` function in prop.

Additionally:
* fixed existing `WfsSearch` component by replacing of deprecated `mode=combobox` property of `Select` by `<AutoComplete>`
* Passing an optional callback function `onBeforeSearch` to be able to manipulate user input before search starts
* Moved `getCombinedRequests` function to `ol-util` 

Note: https://github.com/terrestris/ol-util/pull/30 must be merged and released before.

Please review  @terrestris/devs

### CODE REVIEW

File: src/Field/WfsSearch/WfsSearch.jsx [hunk 1]
```diff
@@ -164,6 +166,13 @@ export class WfsSearch extends React.Component {
      * @type {function}
      */
     onChange: PropTypes.func,
+    /**
+      * Optional callback function, that will be called before WFS search starts.
+      * Can be useful if input value manipulation is needed (e.g. umlaut
+      * replacement `ä => oa` etc.)
```
Comment (weskamm) [inline src/Field/WfsSearch/WfsSearch.jsx]: I think this should read `ae`

**Comment by weskamm** [comment]
Nice PR, did not find any issues. I am missing the `SearchResults` Component, could you add it to this PR or open up another one for it? I would really like to have this included in react-geo.
:+1:

File: src/Field/WfsSearch/WfsSearch.jsx (hunk 1 above)
Comment (annarieger) [reply src/Field/WfsSearch/WfsSearch.jsx]: Fixed in dc65501

**Comment by annarieger** [comment]
@weskamm 
For now I've added an example showing search results in `AgFeatureGrid` component only.

I've opened the [issue](#881) regarding `SearchResultsPanel` component. I would prefer to overhaul it once again before adding to library since this contains some project related comments and code blocks.

**Comment by weskamm** [comment]
:+1:

File: src/Field/WfsSearchInput/WfsSearchInput.jsx [hunk 2]
```diff
@@ -0,0 +1,442 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import {
+  Input,
+  Icon,
```
Comment (dnlkoch) [inline src/Field/WfsSearchInput/WfsSearchInput.jsx]: Minor: Dangling comma.

File: src/Field/WfsSearchInput/WfsSearchInput.jsx [hunk 3]
```diff
@@ -0,0 +1,442 @@
...
+    featurePrefix: PropTypes.string,
+    /**
+     * The feature type names. Required.
+     * @type {String[]}
+     */
+    featureTypes: PropTypes.arrayOf(PropTypes.string).isRequired,
+    /**
+     * SRS name. No srsName attribute will be set on geometries when this is not
+     * provided.
+     * @type {string}
```
Comment (dnlkoch) [inline src/Field/WfsSearchInput/WfsSearchInput.jsx]: `string` -> `String`.

File: src/Field/WfsSearchInput/WfsSearchInput.jsx [hunk 4]
```diff
@@ -0,0 +1,442 @@
...
+     */
+    srsName: PropTypes.string,
+    /**
+     * The output format of the response.
+     * @type {string}
```
Comment (dnlkoch) [inline src/Field/WfsSearchInput/WfsSearchInput.jsx]: `string` -> `String`.

File: src/Field/WfsSearchInput/WfsSearchInput.jsx [hunk 5]
```diff
@@ -0,0 +1,442 @@
...
+     */
+    outputFormat: PropTypes.string,
+    /**
+     * Maximum number of features to fetch.
+     * @type {number}
```
Comment (dnlkoch) [inline src/Field/WfsSearchInput/WfsSearchInput.jsx]: `number` -> `Number`.

File: src/Field/WfsSearchInput/WfsSearchInput.jsx [hunk 6]
```diff
@@ -0,0 +1,442 @@
...
+     */
+    maxFeatures: PropTypes.number,
+    /**
+     * Geometry name to use in a BBOX filter.
+     * @type {string}
```
Comment (dnlkoch) [inline src/Field/WfsSearchInput/WfsSearchInput.jsx]: `string` -> `String`.

File: src/Field/WfsSearchInput/WfsSearchInput.jsx [hunk 7]
```diff
@@ -0,0 +1,442 @@
...
+     */
+    geometryName: PropTypes.string,
+    /**
+     * Optional list of property names to serialize.
+     * @type {String[]}
+     */
+    propertyNames: PropTypes.arrayOf(PropTypes.string),
+    /**
+     * Filter condition. See https://openlayers.org/en/latest/apidoc/ol.format.filter.html
```
Comment (dnlkoch) [inline src/Field/WfsSearchInput/WfsSearchInput.jsx]: Link is outdated. Please make use of [https://openlayers.org/en/latest/apidoc/module-ol_format_filter.html](https://openlayers.org/en/latest/apidoc/module-ol_format_filter.html).

File: src/Field/WfsSearchInput/WfsSearchInput.jsx [hunk 8]
```diff
@@ -0,0 +1,442 @@
...
+     * for more information.
+     * @type {object}
```
Comment (dnlkoch) [inline src/Field/WfsSearchInput/WfsSearchInput.jsx]: `object` -> `Object`.

File: src/Field/WfsSearchInput/WfsSearchInput.jsx [hunk 9]
```diff
@@ -0,0 +1,442 @@
...
+     */
+    filter: PropTypes.object,
+    /**
+     * The ol.map to interact with on selection.
+     *
+     * @type {Object}
```
Comment (dnlkoch) [inline src/Field/WfsSearchInput/WfsSearchInput.jsx]: `Object` -> `ol.Map`.

File: src/Field/WfsSearchInput/WfsSearchInput.jsx [hunk 10]
```diff
@@ -0,0 +1,442 @@
...
+     * The minimal amount of characters entered in the input to start a search.
+     * @type {Number}
+     */
+    minChars: PropTypes.number,
+    /**
+     * An onFetchSuccess callback function which gets called with the
+     * successfully fetched data.
+     * Please note: if omitted only data fetch will be performed and no data
+     * will be shown afterwards!
+     * @type {function}
```
Comment (dnlkoch) [inline src/Field/WfsSearchInput/WfsSearchInput.jsx]: `function` -> `Function`.

File: src/Field/WfsSearchInput/WfsSearchInput.jsx [hunk 11]
```diff
@@ -0,0 +1,442 @@
...
+     */
+    onFetchSuccess: PropTypes.func,
+    /**
+     * An onFetchError callback function which gets called if data fetch is
+     * failed.
+     * @type {function}
```
Comment (dnlkoch) [inline src/Field/WfsSearchInput/WfsSearchInput.jsx]: `function` -> `Function`.

File: src/Field/WfsSearchInput/WfsSearchInput.jsx [hunk 12]
```diff
@@ -0,0 +1,442 @@
...
+     */
+    onFetchError: PropTypes.func,
+    /**
+     * Optional callback function, that will be called if 'clear' button of
+     * input field was clicked.
+     * @type {function}
```
Comment (dnlkoch) [inline src/Field/WfsSearchInput/WfsSearchInput.jsx]: `function` -> `Function`.

File: src/Field/WfsSearchInput/WfsSearchInput.jsx [hunk 13]
```diff
@@ -0,0 +1,442 @@
...
+     */
+    onClearClick: PropTypes.func,
+    /**
+      * Optional callback function, that will be called before WFS search starts.
+      * Can be useful if input value manipulation is needed before (e.g. umlaut
+      * replacement `ä => ae` etc.)
+      * @type {function}
```
Comment (dnlkoch) [inline src/Field/WfsSearchInput/WfsSearchInput.jsx]: `function` -> `Function`.

File: src/Field/WfsSearchInput/WfsSearchInput.jsx [hunk 14]
```diff
@@ -0,0 +1,442 @@
...
+      */
+    onBeforeSearch: PropTypes.func,
+    /**
+     * Options which are added to the fetch-POST-request. credentials is set to
+     * 'same-origin' as default but can be overwritten. See also
+     * https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/fetch
+     * @type {object}
```
Comment (dnlkoch) [inline src/Field/WfsSearchInput/WfsSearchInput.jsx]: `object` -> `Object`.

File: src/Field/WfsSearchInput/WfsSearchInput.jsx [hunk 15]
```diff
@@ -0,0 +1,442 @@
...
+     */
+    additionalFetchOptions: PropTypes.object,
+    /**
+     * Options which are passed to the constructor of the ol.format.WFS.
+     * compare: https://openlayers.org/en/latest/apidoc/ol.format.WFS.html
+     * @type {object}
```
Comment (dnlkoch) [inline src/Field/WfsSearchInput/WfsSearchInput.jsx]: `object` -> `Object`.

File: src/Field/WfsSearchInput/WfsSearchInput.jsx [hunk 16]
```diff
@@ -0,0 +1,442 @@
...
+     */
+    wfsFormatOptions: PropTypes.object,
+    /**
+     * Delay in ms before actually sending requests.
+     * @type {number}
```
Comment (dnlkoch) [inline src/Field/WfsSearchInput/WfsSearchInput.jsx]: `number` -> `Number`.

File: src/Field/WfsSearchInput/WfsSearchInput.jsx (hunk 15 above)
Comment (dnlkoch) [reply src/Field/WfsSearchInput/WfsSearchInput.jsx]: Link is outdated, please make use of [https://openlayers.org/en/latest/apidoc/module-ol_format_WFS.html](https://openlayers.org/en/latest/apidoc/module-ol_format_WFS.html).

File: src/Field/WfsSearchInput/WfsSearchInput.jsx [hunk 17]
```diff
@@ -0,0 +1,442 @@
...
+    delay: PropTypes.number
+  }
+
+  static defaultProps = {
+    srsName: 'EPSG:3857',
+    outputFormat: 'application/json',
+    minChars: 3,
+    additionalFetchOptions: {},
+    attributeDetails: {},
+    delay: 300,
```
Comment (dnlkoch) [inline src/Field/WfsSearchInput/WfsSearchInput.jsx]: Dangling comma.

File: src/Field/WfsSearchInput/WfsSearchInput.jsx [hunk 18]
```diff
@@ -0,0 +1,442 @@
...
+   */
+  onUpdateInput(evt) {
+    const {
+      minChars,
+      onBeforeSearch
+    } = this.props;
+
+    this.setState({
+      data: []
+    });
```
Comment (dnlkoch) [inline src/Field/WfsSearchInput/WfsSearchInput.jsx]: This call may be async.

File: src/Field/WfsSearchInput/WfsSearchInput.jsx [hunk 19]
```diff
@@ -0,0 +1,442 @@
...
+   *
+   * @param {String} error The errorstring.
+   */
+  onFetchError(error) {
+    const {
+      onFetchError
+    } = this.props;
+    Logger.error(`Error while requesting WFS GetFeature: ${error}`);
+    this.setState({
+      fetching: false
```
Comment (dnlkoch) [inline src/Field/WfsSearchInput/WfsSearchInput.jsx]: Unsure, but do we want to reset `state.data` here?

File: src/Field/WfsSearchInput/WfsSearchInput.jsx [hunk 20]
```diff
@@ -0,0 +1,442 @@
...
+      onClearClick
+    } = this.props;
+    this.inputRef.input.value = '';
+    this.setState({
+      data: []
+    });
+
+    if (isFunction(onClearClick)) {
+      onClearClick();
+    }
```
Comment (dnlkoch) [inline src/Field/WfsSearchInput/WfsSearchInput.jsx]: Please move custom callback to `setState()` callback.

**Comment by annarieger** [comment]
O, bad timing @dkoch, I've just merged it without taking in account your comments.. A follow-up is already coming :see_no_evil:

**Pull request merged**